pkg/bytecode/vm_test.go

```diff
@@ -89,94 +85,78 @@ func TestUserError(t *testing.T) {
 func TestBoolExpressions(t *testing.T) {
 	tests := []testCase{
 		{
-			name: "literal true",
-			input: `
-			x := true
-			x = x
-			`,
+			name:              "literal true",
+			input:             "x := true",
 			expectedStackTop:  true,
 			expectedConstants: []any{},
 			expectedInstructions: []Instructions{
 				mustMake(t, OpTrue),
 				mustMake(t, OpSetGlobal, 0),
-				mustMake(t, OpGetGlobal, 0),
-				mustMake(t, OpSetGlobal, 0),
 			},
 		},
 		{
-			name: "literal false",
-			input: `
-			x := false
-			x = x
-			`,
+			name:              "literal false",
+			input:             "x := false",
 			expectedStackTop:  false,
 			expectedConstants: []any{},
 			expectedInstructions: []Instructions{
 				mustMake(t, OpFalse),
 				mustMake(t, OpSetGlobal, 0),
-				mustMake(t, OpGetGlobal, 0),
-				mustMake(t, OpSetGlobal, 0),
 			},
 		},
 		{
-			name: "not operator",
-			input: `
-			x := !true
-			x = x
-			`,
+			name:              "not operator",
+			input:             "x := !true",
 			expectedStackTop:  false,
 			expectedConstants: []any{},
 			expectedInstructions: []Instructions{
 				mustMake(t, OpTrue),
 				mustMake(t, OpNot),
 				mustMake(t, OpSetGlobal, 0),
-				mustMake(t, OpGetGlobal, 0),
-				mustMake(t, OpSetGlobal, 0),
 			},
 		},
 		{
-			name: "equal operator",
-			input: `
-			x := 1 == 1
-			x = x
-			`,
+			name:              "equal operator",
+			input:             "x := 1 == 1",
 			expectedStackTop:  true,
 			expectedConstants: []any{1, 1},
 			expectedInstructions: []Instructions{
 				mustMake(t, OpConstant, 0),
 				mustMake(t, OpConstant, 1),
 				mustMake(t, OpEqual),
 				mustMake(t, OpSetGlobal, 0),
-				mustMake(t, OpGetGlobal, 0),
-				mustMake(t, OpSetGlobal, 0),
 			},
 		},
 		{
-			name: "not operator",
-			input: `
-			x := 1 != 1
-			x = x
-			`,
+			name:              "not operator",
+			input:             "x := 1 != 1",
 			expectedStackTop:  false,
 			expectedConstants: []any{1, 1},
 			expectedInstructions: []Instructions{
 				mustMake(t, OpConstant, 0),
 				mustMake(t, OpConstant, 1),
 				mustMake(t, OpNotEqual),
 				mustMake(t, OpSetGlobal, 0),
-				mustMake(t, OpGetGlobal, 0),
-				mustMake(t, OpSetGlobal, 0),
 			},
 		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			tt.input += "\nx = x"
 			program, err := parser.Parse(tt.input, parser.Builtins{})
 			assert.NoError(t, err, "parser error")
 			comp := NewCompiler()
 			err = comp.Compile(program)
 			assert.NoError(t, err, "compiler error")
-			vm := NewVM(comp.Bytecode())
+			bytecode := comp.Bytecode()
+			// add expectation of x = x as the last instructions
+			tt.expectedInstructions = append(tt.expectedInstructions,
```

Comment:
Similar to before - best to keep you test cases constant. In this case I think I'd just subslice the result:

```
bci := bytecode.Instructions
assertInstructions(t, tt.expectedInstructions, bci[:len(bci)-2])
```

The reason I prefer this is if something breaks such that `x=x` generates different opcodes (perhaps the opcode operand is a different global and not number 0), the test wont fail for reasons unrelated to the specific test. Hopefully the test for the assignment will fail so it points at the buggy code, but it should mean all the tests don't start failing. It's not foolproof as the code generated for `x = x` may emit more or less than 2 instructions, but that's a narrower set of failures.